stl/src/vector_algorithms.cpp

```diff
@@ -451,6 +452,358 @@ __declspec(noalias) void __cdecl __std_reverse_copy_trivially_copyable_8(
+        unsigned _Bingo = static_cast<unsigned>(_mm256_movemask_epi8(_Traits::_Cmp_avx(_Data, _Comparand)));
+
+        if ((_Bingo &= _Mask) != 0) {
+            unsigned long _Offset = _tzcnt_u32(_Bingo);
+            _Advance_bytes(_First, _Offset);
+            return _First;
+        }
+
+        for (;;) {
+            _Data  = _mm256_load_si256(static_cast<const __m256i*>(_First));
+            _Bingo = static_cast<unsigned>(_mm256_movemask_epi8(_Traits::_Cmp_avx(_Data, _Comparand)));
+
+            if (_Bingo != 0) {
+                unsigned long _Offset = _tzcnt_u32(_Bingo);
+                _Advance_bytes(_First, _Offset);
+                return _First;
+            }
+
+            _Advance_bytes(_First, 32);
+        };
+    }
+    if (_Traits::_Sse_available()) {
+        // We read by vector-sized pieces, and we align pointers to vector-sized boundary.
+        // From start partial piece we mask out matches that don't belong to the range.
+        // This makes sure we never cross page boundary, thus we read 'as if' sequentially.
+        constexpr size_t _Vector_pad_mask = 0xF;
+        constexpr unsigned _Full_mask     = 0xFFFF;
+
+        const __m128i _Comparand  = _Traits::_Set_sse(_Val);
+        const intptr_t _Pad_start = reinterpret_cast<intptr_t>(_First) & _Vector_pad_mask;
+        const unsigned _Mask      = (_Full_mask << _Pad_start);
+        _Advance_bytes(_First, -_Pad_start);
+
+        __m128i _Data   = _mm_load_si128(static_cast<const __m128i*>(_First));
+        unsigned _Bingo = static_cast<unsigned>(_mm_movemask_epi8(_Traits::_Cmp_sse(_Data, _Comparand)));
+
+        if ((_Bingo &= _Mask) != 0) {
+            unsigned long _Offset;
+            _BitScanForward(&_Offset, _Bingo);
+            _Advance_bytes(_First, _Offset);
+            return _First;
+        }
+
+        for (;;) {
+            _Data  = _mm_load_si128(static_cast<const __m128i*>(_First));
+            _Bingo = static_cast<unsigned>(_mm_movemask_epi8(_Traits::_Cmp_sse(_Data, _Comparand)));
+
+            if (_Bingo != 0) {
+                unsigned long _Offset;
+                _BitScanForward(&_Offset, _Bingo);
+                _Advance_bytes(_First, _Offset);
+                return _First;
+            }
+
+            _Advance_bytes(_First, 16);
+        };
+    }
+
+    return _Find_trivial_unsized_fallback(_First, _Val);
+}
+
+
+template <class _Traits, class _Ty>
+const void* __stdcall __std_find_trivial(const void* _First, const void* _Last, _Ty _Val) noexcept {
+    size_t Size_bytes = _Byte_length(_First, _Last);
+
+    const size_t _Avx_size = Size_bytes & ~size_t{0x1F};
+    if (_Avx_size != 0 && _Use_avx2()) {
+        const __m256i _Comparand = _Traits::_Set_avx(_Val);
+        const void* _Stop_at     = _First;
+        _Advance_bytes(_Stop_at, _Avx_size);
+        do {
+            const __m256i _Data = _mm256_loadu_si256(static_cast<const __m256i*>(_First));

```

Comment:
I'm afraid I don't understand the question.

For all algorithms here, the buffer isn't assumed to be 32 or 16 bytes aligned, so aligned instruction may fail.

The only exception is unsized algorithms. They align the buffer before processing to avoid hard page fault, so they can rely on alignment.

Can use unalignned instruction for unsized algorithms too. But here we must use unaligned.

Reply:
Ah, I think I got it. Note that the usual loop is compiled anyway. `if constexpr (_Vector_alg_in_find_is_safe<_InIt, _Ty>)` don't even have `else`. So the usual loop checks types.